tmt/lint.py

```diff
@@ -0,0 +1,308 @@
+# adding color when it shouldn't have.
+#
+# _OUTCOME_TO_MARK = {
+#     LinterOutcome.SKIP: '-',
+#     LinterOutcome.PASS: '\N{heavy check mark}',
+#     LinterOutcome.FAIL: '\N{heavy multiplication x}',
+#     LinterOutcome.WARN: '\N{exclamation mark}',
+#     LinterOutcome.FIXED: '\N{heavy check mark}'
+# }
+
+_OUTCOME_TO_MARK = {
+    LinterOutcome.SKIP: 'skip',
+    LinterOutcome.PASS: 'pass',
+    LinterOutcome.FAIL: 'fail',
+    LinterOutcome.WARN: 'warn',
+    LinterOutcome.FIXED: 'fix '
+    }
+
+_OUTCOME_TO_COLOR = {
+    LinterOutcome.SKIP: 'green',
+    LinterOutcome.PASS: 'green',
+    LinterOutcome.FAIL: 'red',
+    LinterOutcome.WARN: 'yellow',
+    LinterOutcome.FIXED: 'green'
+    }
+
+
+#: Info on how a linter decided: linter itself, its outcome & the message.
+LinterRuling = Tuple['Linter', LinterOutcome, LinterOutcome, str]
+#: A return value type of a single linter.
+LinterReturn = Generator[Tuple[LinterOutcome, str], None, None]
+#: A linter itself, a callable method.
+LinterCallback = Callable[['Lintable'], LinterReturn]
+
+
+@dataclasses.dataclass(init=False)
+class Linter:
+    """ A single linter """
+
+    callback: LinterCallback
+    id: str
+
+    def __init__(self, callback: LinterCallback) -> None:
+        self.callback = callback
+
+        if not callback.__doc__:
+            raise tmt.utils.GeneralError(f"Linter '{callback}' lacks docstring.")
+
+        id, _ = textwrap.dedent(callback.__doc__).strip().split(':', 1)
+
+        self.id = id.strip()
+
+
+class Lintable(Generic[LintableT]):
+    """ Mixin class adding support for linting of class instances """
+
+    # Declare linter registry as a class variable, but do not initialize it. If initialized
+    # here, the mapping would be shared by all classes, which is not a desirable attribute.
+    # Instead, mapping will be created by `get_linter_registry()`.
+    _linter_registry: ClassVar[List[Linter]]
+
+    # Keep this method around, to correctly support Python's method resolution order.
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+    # Cannot use @property as this must remain classmethod
+    @classmethod
+    def get_linter_registry(cls) -> List[Linter]:
+        """ Return - or initialize - linter registry """
+
+        if not hasattr(cls, '_linter_registry'):
+            cls._linter_registry = []
+
+        return cls._linter_registry
+
+    @classmethod
+    def discover_linters(cls) -> None:
+        """
+        Discover and register all linters implemented by this class.
+
+        A linter is a method whose name starts with ``lint_`` prefix. It must
+        have a docstring which serves as a hint for ``--list-checks`` output.
+        """
+
+        for name in dir(cls):
+            if not name.startswith('lint_'):
+                continue
+
+            cls.get_linter_registry().append(tmt.lint.Linter(getattr(cls, name)))
+
+    @classmethod
+    def resolve_enabled_linters(
```

Comment:
It would be great if we could store enabled, or better yet disabled, checks in some config, e.g. in .fmf/something
basically the same thing we are doing for flake, mypy and now ruff
completely fine to do in another patch